core/src/impl/Kokkos_GraphImpl.hpp

```diff
@@ -31,6 +31,10 @@
 namespace Kokkos {
 namespace Impl {
 
+template <typename T>
+struct is_graph_then<T, std::enable_if_t<Kokkos::Impl::is_specialization_of_v<
+                            T, GraphNodeThenImpl>>> : public std::true_type {};
```

Comment:
A related issue may be to think about a convention for organizing/naming/... new node types and how the concepts/constraints/... should behave.

Currently, we have the `GraphNodeImpl<ExecutionSpace, Kernel, Predecessor>` that holds the `m_kernel` member of type `Kernel`. And, currently, this `Kernel` is an instance of `GraphNodeKernelImpl<ExecutionSpace, PolicyType, Functor, PatternTag, ...>`. It can also be an instance of `GraphNodeAggregateKernelDefaultImpl<ExecutionSpace>` for the default implementation and of similar classes for non-default backends. The concept `Kokkos::Impl::is_graph_kernel` is `true` for all such instances, i.e., currently, it is `true` for aggregate nodes too.

With the proposed change to add the "then", we'll have:
- `GraphNodeKernelImpl<ExecutionSpace, PolicyType, Functor, PatternTag, ...>`
- `GraphNodeAggregateKernelDefaultImpl<ExecutionSpace>` and similar classes for non-default backends 
- `GraphNodeThenImpl<ExecutionSpace, Functor>`

And, with the proposed change, would not be `true` for the "then" node, if I understand correctly. The concepts `is_graph_kernel` and `is_graph_then` are also implemented differently, i.e., `is_graph_kernel` checks for the presence of a member type alias, and `is_graph_then` uses the `is_specialization_of_v`.

So, these are a few thoughts:
- How to get consistency in the naming? E.g. "Kernel" in the name of the aggregate node, but not in the name of the "then" node. One option may be to rename `GraphNodeAggregateKernelDefaultImpl` as `GraphNodeAggregateDefaultImpl`. With similar renaming for the non-default backends.
- What should `is_graph_kernel` do? Should it be true for an aggregate node? One option may be to define all such concepts like it is proposed now for the `is_graph_then`, i.e. with `is_specialization_of_v`. So `is_graph_kernel` would check whether the type is an instance of `GraphNodeKernelImpl`, and it would evaluate to false for the case of an aggregate node. Such a change would likely require some modifications to constraints/asserts in the code.